File: backend/app/core/response.py
"""المغلف الموحد للاستجابات — المصدر الوحيد لشكل الرد."""
import uuid
from typing import Any

from fastapi import Request
from fastapi.responses import JSONResponse


def ok(data: Any = None, request_id: str | None = None) -> dict:
    return {"ok": True, "data": data, "meta": {"requestId": request_id}}


def fail(code: str, message: str, request_id: str | None = None, status: int = 500) -> JSONResponse:
    return JSONResponse(
        status_code=status,
        content={
            "ok": False,
            "error": {"code": code, "message": message},
            "meta": {"requestId": request_id},
        },
    )


def get_request_id(request: Request | None) -> str | None:
    if request is None:
        return None
    return getattr(request.state, "request_id", request.headers.get("X-Request-Id"))


async def request_id_middleware(request: Request, call_next):
    request_id = request.headers.get("X-Request-Id") or uuid.uuid4().hex[:12]
    request.state.request_id = request_id
    response = await call_next(request)
    response.headers["X-Request-Id"] = request_id
    return response

File: realtime/src/common/response.ts
/** المغلف الموحد — نسخة TypeScript من backend/app/core/response.py
 * نجاح: { ok:true, data, meta:{requestId} }
 * فشل:  { ok:false, error:{code,message}, meta:{requestId} }
 * الفرونت يفك data فقط عندما ok===true.
 */
export interface ApiMeta {
  requestId: string | null;
}

export interface ApiSuccess<T = unknown> {
  ok: true;
  data: T;
  meta: ApiMeta;
}

export interface ApiFailure {
  ok: false;
  error: { code: string; message: string };
  meta: ApiMeta;
}

export type ApiEnvelope<T = unknown> = ApiSuccess<T> | ApiFailure;

export function ok<T>(data: T, requestId: string | null = null): ApiSuccess<T> {
  return { ok: true, data, meta: { requestId } };
}

export function fail(
  code: string,
  message: string,
  requestId: string | null = null,
): ApiFailure {
  return { ok: false, error: { code, message }, meta: { requestId } };
}

/** هل الرد مغلف جاهز من بايثون؟ لا نعيد تغليفه. */
export function isEnvelope<T = unknown>(value: unknown): value is ApiEnvelope<T> {
  if (typeof value !== "object" || value === null) return false;
  return (value as Record<string, unknown>).ok === true || (value as Record<string, unknown>).ok === false;
}
